src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala

```diff
@@ -636,6 +637,62 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
       } }
   }
 
+  override def addCallFailureEvent(workflowId: WorkflowId, executionKey: ExecutionDatabaseKey,
+                                           failure: FailureEventEntry)(implicit ec: ExecutionContext): Future[Unit] = {
+    val action = for {
+      workflowExecution <- dataAccess.workflowExecutionsByWorkflowExecutionUuid(workflowId.toString).result.head
+      execution <- executionKey match {
+        case ExecutionDatabaseKey(callFqn, Some(index), attempt) => dataAccess.executionsByWorkflowExecutionUuidAndCallFqnAndShardIndexAndAttempt(workflowId.toString, callFqn, index, attempt).result.head
+        case ExecutionDatabaseKey(callFqn, None, attempt) => dataAccess.executionsByWorkflowExecutionUuidAndCallFqnAndAttempt(workflowId.toString, callFqn, attempt).result.head
+      }
+
+      _ <- dataAccess.failureEventsAutoInc +=
+        new FailureEvent(
+          workflowExecution.workflowExecutionId.get,
+          Option(execution.executionId.get),
+          StringUtils.abbreviate(failure.failure, 255),
+          new Timestamp(failure.timestamp.getMillis))
+    } yield ()
+
+    runTransaction(action)
+  }
+
+  override def addWorkflowFailureEvent(workflowId: WorkflowId, failure: FailureEventEntry)(implicit ec: ExecutionContext): Future[Unit] = {
+    val action = for {
+      workflowExecution <- dataAccess.workflowExecutionsByWorkflowExecutionUuid(workflowId.toString).result.head
+
+      _ <- dataAccess.failureEventsAutoInc +=
+        new FailureEvent(
+          workflowExecution.workflowExecutionId.get,
+          None,
+          StringUtils.abbreviate(failure.failure, 255),
+          new Timestamp(failure.timestamp.getMillis))
+    } yield ()
+
+    runTransaction(action)
+  }
+
+  private def toFailureEntries(failures: Traversable[(String, FailureEvent)])(implicit ec: ExecutionContext): Seq[Future[QualifiedFailureEventEntry]] = {
+    (failures map { case (workflowId, failure) =>
+      val failureMessage = failure.failure
+      val failureTimestamp = new DateTime(failure.timestamp.getTime)
+      failure.executionId match {
+        case None => Future.successful(QualifiedFailureEventEntry(workflowId, None, failureMessage, failureTimestamp))
+        case Some(executionId) =>
+          val executions = runTransaction(dataAccess.executionsByExecutionId(executionId).result.head)
+          executions map { execution => {
+            val executionDatabaseKey = ExecutionDatabaseKey(execution.callFqn, execution.index.toIndex, execution.attempt)
+            QualifiedFailureEventEntry(workflowId, Some(executionDatabaseKey), failureMessage, failureTimestamp)
+          }}
+      }
+    }).toSeq
+  }
+
+  def getFailureEvents(workflowId: WorkflowId)(implicit ec: ExecutionContext): Future[Seq[QualifiedFailureEventEntry]] = {
+    val action = dataAccess.failuresByWorkflowExecutionUuid(workflowId.toString).result
+    runTransaction(action) flatMap { x => Future.sequence(toFailureEntries(x)) }
```

Comment:
Good catch, and actually `toFailureEntries` is running transactions in a loop.  :frowning:  Let me see what I can do about this.